src/dma.md

```diff
@@ -0,0 +1,348 @@
+# Direct Memory Access (DMA)
+
+This section covers the core requirements for building a memory safe API around
+DMA transfers.
+
+The DMA peripheral is used to perform memory transfers in parallel to the work
+of the processor (the execution of the main program). A DMA transfer is more or
+less equivalent to spawning a thread (see [`thread::spawn`]) to do a `memcpy`.
+We'll use the fork-join model to illustrate the requirements of a memory safe
+API.
+
+[`thread::spawn`]: https://doc.rust-lang.org/std/thread/fn.spawn.html
+
+Consider the following DMA primitives:
+
+``` rust
+{{#include ../ci/dma/src/lib.rs:6:57}}
+{{#include ../ci/dma/src/lib.rs:59:60}}
+```
+
+Assume that the `Dma1Channel1` is statically configured to work with serial port
+(AKA UART or USART) #1: `Serial1`. `Serial1` provides the following *blocking*
+API:
+
+``` rust
+{{#include ../ci/dma/src/lib.rs:62:72}}
+{{#include ../ci/dma/src/lib.rs:74:80}}
+{{#include ../ci/dma/src/lib.rs:82:83}}
+```
+
+Let's say we want to extend `Serial1` API to (a) asynchronously send out a
+buffer and (b) asynchronously fill a buffer.
+
+We'll start with a memory unsafe API and we'll iterate on it until it's
+completely memory safe. On each step we'll show you how the API can be broken to
+make you aware of the issues that need to be addressed when dealing with
+asynchronous memory operations.
+
+## A first stab
```

Comment:
Can't the `Transfer` never finish a transaction? For example, if a DMA is setup into continuous circular mode.
Should we think about how to handle these kind of uses as well? Where one just "sets it up" and never touches it again.
Or maybe this is covered and I'm missing it?

Reply:
> Can't the Transfer never finish a transaction?

No, the channel is configured in "one-shot" mode. I should note that explicitly in the text.

> Should we think about how to handle these kind of uses as well?

I'd like to have the community experiment with circular mode as there are a few options here in terms of API design, and I worry that if I write down one option here people will feel that they don't need to try something different.